modules/javafx.graphics/src/main/native-iio/jpegloader.c

```diff
@@ -1606,50 +1620,49 @@ JNIEXPORT jboolean JNICALL Java_com_sun_javafx_iio_jpeg_JPEGImageLoader_decompre
     struct jpeg_source_mgr *src = cinfo->src;
     sun_jpeg_error_ptr jerr;
     int bytes_per_row = cinfo->output_width * cinfo->output_components;
-    JSAMPROW scanline_ptr = (JSAMPROW) malloc(bytes_per_row * sizeof (JSAMPLE));
     int offset = 0;
 
+    JSAMPROW scanline_ptr = (JSAMPROW) malloc(bytes_per_row * sizeof (JSAMPLE));
```

Comment:
You can remove quite a few calls to `free` if you move the memory allocation for `scanline_ptr` just [before it's usage](https://github.com/openjdk/jfx/blob/7af932b7f5215949776ec79fb2a5484c521b21a1/modules/javafx.graphics/src/main/native-iio/jpegloader.c#L1690). Also free it as soon as you are done with it.

Reply:
PR is updated according to this comment, please have a look.

Reply:
> In general, this makes sense. I need to look into more detail that the additional calls for freeing resources (in case of errors) don't cause e.g. segmentation violations and lead to a crash -- which would be worse than throwing an Exception.

The native code throws two Exceptions in case of error, 
1. OutOfMemory : This would exit the application.
2. IOException: There is no action on this exception for now. Only callstack is printed when -Dprism.verbose=true.

> I expect memory consumption to be similar before and after this patch if you don't run into errors, but did you check memory consumption before/after this patch in case of errors?
I verified this now, by changing native code to always throw an exception from different error cases. The memory is freed correctly and remains in same range for any number of images,